core/host.go

```diff
@@ -5,122 +5,54 @@ import (
 	"fmt"
 	"path/filepath"
 	"strings"
-	"sync"
 
-	bkclient "github.com/moby/buildkit/client"
 	"github.com/moby/buildkit/client/llb"
-	bkgw "github.com/moby/buildkit/frontend/gateway/client"
-	"github.com/moby/buildkit/solver/pb"
 	specs "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
+type Host struct {
+	Workdir string
+}
+
+func NewHost(workdir string) *Host {
+	return &Host{
+		Workdir: workdir,
+	}
+}
+
 type HostVariable struct {
 	Name string `json:"name"`
 }
 
-type HostDirectory struct {
-	ID HostDirectoryID `json:"id"`
-}
+func (host *Host) Directory(ctx context.Context, dirPath string, platform specs.Platform) (*Directory, error) {
+	var absPath string
+	var err error
+	if filepath.IsAbs(dirPath) {
+		absPath = dirPath
+	} else {
+		absPath = filepath.Join(host.Workdir, dirPath)
 
-func (dir *HostDirectory) Read(ctx context.Context, platform specs.Platform) (*Directory, error) {
-	id := string(dir.ID)
+		absPath, err = filepath.EvalSymlinks(absPath)
+		if err != nil {
+			return nil, fmt.Errorf("eval symlinks: %w", err)
+		}
+
+		if !strings.HasPrefix(absPath, host.Workdir) {
+			return nil, fmt.Errorf("path %q escapes workdir; use an absolute path instead", dirPath)
+		}
+	}
 
 	// copy to scratch to avoid making buildkit's snapshot of the local dir immutable,
 	// which makes it unable to reused, which in turn creates cache invalidations
 	// TODO: this should be optional, the above issue can also be avoided w/ readonly
 	// mount when possible
 	st := llb.Scratch().File(llb.Copy(llb.Local(
-		id,
-		// TODO: better shared key hint?
-		llb.SharedKeyHint(id),
+		absPath, // TODO: is there a better thing to set here? ...seems ok?
+		llb.SharedKeyHint(absPath),
+		llb.LocalUniqueID(absPath),
```

Comment:
Do you know what LocalUniqueID does? I've never seen it actually be used, only set (source attr def [here](https://github.com/sipsma/buildkit/blob/068cf686a7e5c3254244d0acc54636f4f393611e/solver/pb/attr.go#L10-L10)).

My best guess is that it just ends up in the LLB digest, which then results in the solver not deduplicating it when the session ID is not included. If that's the case though, then it's probably not technically needed here since absPath is already in the digest. It doesn't cause any harm though either so not a big deal, just curious if you knew anything more.

Reply:
The only reason I'm setting it is to make the generated LLB stable so I can write tests for invariants. I had a comment for this in another branch:

```go
        localOpts := []llb.LocalOption{
                // synchronize concurrent filesyncs for the same path
                llb.SharedKeyHint(localID),

                // make the LLB stable so we can test invariants like:
                //
                //   workdir == directory(".")
                llb.LocalUniqueID(localID),
        }
```

Probably worth bringing back for clarity.

Reply:
Ah I see the line where it defaults to a random ID if not set now, makes sense!